src/vsr/journal.zig

```diff
@@ -669,117 +812,541 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+                        // This is tricky because the redundant headers are written in batches.
+                        self.set_header_as_dirty(prepare);
+                    }
+                    self.faulty.clear(slot);
+                },
+                .vsr => {
+                    // Faulty headers are loaded as reserved to prevent unsafe batch updates of the
+                    // redundant headers. See `recover_prepares_callback()` for more detail.
+                    self.headers[slot.index] = Header.reserved(replica.cluster, slot.index);
+                    self.dirty.set(slot);
+                    self.faulty.set(slot);
+                },
+            }
+
+            switch (case.decision) {
+                .eql, .nil => {
+                    log.debug("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
+                .fix, .vsr => {
+                    log.warn("{}: recover_prepares: recovered slot={} label={s} decision={s}", .{
+                        self.replica,
+                        slot.index,
+                        case.label,
+                        @tagName(case.decision),
+                    });
+                },
             }
 
-            // We must release before we call `recover_headers()` in case Storage is synchronous.
-            // Otherwise, we would run out of messages and reads.
             replica.message_bus.unref(read.message);
             self.reads.release(read);
+            self.recover_prepares(Slot{ .index = slot.index + 1 });
+        }
+
+        fn recover_done(self: *Self) void {
+            const replica = @fieldParentPtr(Replica, "journal", self);
+
+            log.debug("{}: recover_done: dirty={} faulty={}", .{
+                self.replica,
+                self.dirty.count,
+                self.faulty.count,
+            });
+
+            assert(self.faulty.count == 0 or replica.replica_count != 1);
+            assert(self.faulty.count >= self.dirty.count);
+            assert(self.faulty.count < slot_count);
+            if (self.headers[0].op == 0 and self.headers[0].command == .prepare) {
+                assert(self.headers[0].checksum == Header.root_prepare(replica.cluster).checksum);
+            }
 
-            self.recover_headers(offset + buffer.len, version);
+            {
+                var min_op: u64 = std.math.maxInt(u64);
+                var max_op: u64 = 0;
+                for (self.headers) |*header, slot| {
+                    if (header.command == .reserved) {
+                        assert(header.cluster == replica.cluster);
+                        assert(header.op == slot);
+                        continue;
+                    }
+
+                    assert(header.command == .prepare);
```

Comment:
> Once the log wraps, if we have a header that's invalid and a prepare that's invalid, what header do we stick in headers?

A `reserved` header. Also, the `op` of the reserved header is always the `slot` index exactly.

> I think this assert could fail, because on line 1304 we have asserted that the log can't have wrapped if the header is `.reserved`.

See above; the `op` of the reserved header is always less than `slot_count`. This is necessary because we don't actually know what the "correct" `op` is for the slot.